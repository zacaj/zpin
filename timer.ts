import { Utils, Opaque, assert } from './util';
import { Events } from './events';
import { StateEvent } from './state';
import { Log } from './log';
import { fork, settleForks } from './promises';

export type Time = Opaque<number, 'Time'>;

export class Timer {
    static startTime = new Date().getTime();

    static mockTime?: number;
    static get time(): Time {
        if (Utils.stateAccessRecorder) {
            Utils.stateAccessRecorder(Timer, 'time');
        }

        if (Timer.mockTime !== undefined) return Timer.mockTime as Time;
        
        return new Date().getTime() - Timer.startTime as Time;
    }

    static queue: TimerQueueEntry[] = [];

    private static addToQueue(entry: TimerQueueEntry) {
        let i = 0;
        for (; i<Timer.queue.length; i++) {
            if (Timer.queue[i].time >= entry.time) {
                break;
            }
        }
        
        Timer.queue.splice(i, 0, entry);
    }

    static schedule(func: TimerCallback, at: Time, context?: string): TimerQueueEntry {
        const entry = { func, time: at, context };
        Timer.addToQueue(entry);
        return entry;
    }

    static callIn(func: TimerCallback, ms: number, context?: string) {
        return Timer.schedule(func, time() + ms as Time, context);
    }

    static setInterval(func: TimerCallback, ms: number, context?: string) {
        const entry = Timer.callIn(func, ms, context);
        entry.repeat = ms as Time;
        return entry;
    }

    static cancel(entry: TimerQueueEntry) {
        Timer.queue.remove(entry);
    }

    static curTime = time();

    static async fireTimers(before = time(), updateMock = false) {
        while (true) {
            const entry = Timer.queue[0];
            if (entry && entry.time <= before) {
                Timer.cancel(entry);
                try {
                    if (updateMock)
                        Timer.mockTime = entry.time;
                    await entry.func(entry);
                    await settleForks();
                } catch (err) {
                    Log.error('console', 'error running entry %o: ', entry, err);
                    debugger;
                }
                if (entry.repeat) {
                    entry.time = time() + entry.repeat as Time;
                    Timer.addToQueue(entry);
                }
            } else {
                break;
            }
        }
        
        if (updateMock)
            Timer.mockTime = before;
    }

    static reset() {
        this.queue = [];
    }
}
setInterval(() => {
    if (Timer.mockTime) return;
    const tim = time();
    Events.fire(new StateEvent(Timer, 'time', tim, Timer.curTime));
    Timer.curTime = tim;
    fork(Timer.fireTimers(tim), 'timer tick');
}, 5);

export type TimerCallback = (entry: TimerQueueEntry) => Promise<any>|any;
export type TimerQueueEntry = {
    time: Time;
    func: TimerCallback;
    repeat?: Time;
    context?: string;
};

export function safeSetInterval(func: () => any, ms: number, context: string) {
    // setInterval(() => {
    //     if (Timer.mockTime) return 'skipped';
    //     return func();
    // }, ms);
    Timer.setInterval(func, ms, context);
}
export function safeSetTimeout(func: () => any, ms: number, context: string) {
    // assert(!Timer.mockTime);
    // setTimeout(func, ms);
    Timer.callIn(func, ms, context);
}
export function time() {
    return Timer.time;
}
export async function setTime(ms?: number) {
    const lastTime = Timer.mockTime;
    Timer.mockTime = ms;
    const newTime = time();
    if (ms === undefined && lastTime) {
        Timer.startTime -= lastTime - newTime;
        const newNewTime = time();
        assert(Math.abs(newNewTime - lastTime) < 2);
        Log.log('console', 'adjusted time to %i', newNewTime);
    }
    await Timer.fireTimers(newTime, !!ms);
    Events.fire(new StateEvent(Timer, 'time', ms as Time ?? time(), lastTime as Time));
    if (Timer.mockTime !== undefined)
        await settleForks();
}
export async function passTime(ms = 1) {
    assert(Timer.mockTime !== undefined);
    await setTime(Timer.mockTime! + ms);
}

export async function wait(ms: number, context?: string) {
    return new Promise(resolve => safeSetTimeout(resolve, ms, context ?? 'wait '+ms));
}