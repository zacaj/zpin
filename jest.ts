import { resetMachine, MachineOutputs, machine, Solenoid } from './machine';
import { resetSwitchMatrix } from './switch-matrix';
import { Events } from './events';
import { MPU } from './mpu';
import { Timer, setTime, wait } from './timer';
import { Log } from './log';

beforeEach(async () => {
    jest.spyOn(Timer, 'schedule').mockRestore();
    jest.spyOn(Events, 'listen').mockRestore();
    jest.spyOn(Events, 'fire').mockRestore();
    jest.spyOn(Timer, 'callIn').mockRestore();
    jest.spyOn(Log, 'init').mockImplementation(() => { throw 'unexpected' });
    jest.spyOn(Log, 'write').mockReturnValue();
    jest.spyOn(Log, 'logMessage').mockReturnValue();
    jest.spyOn(Log, 'trace').mockReturnValue();
    Timer.reset();
    Events.resetAll();
    resetSwitchMatrix();
    resetMachine();
    await setTime(1);
    jest.spyOn(MPU, 'sendCommandCode').mockImplementation(async (cmd) => {
        debugger;
        
        expect(cmd).toBe('mocked');

        return {
            code: 200,
            resp: 'mocked',
        };
    });
});

afterEach(async () => {
    Events.resetAll();
    Timer.reset();
    await setTime();
    jest.spyOn(Timer, 'callIn').mockReturnValue({} as any);
    jest.spyOn(Timer, 'schedule').mockReturnValue({} as any);
    jest.spyOn(Events, 'listen').mockReturnValue({} as any);
    jest.spyOn(Events, 'fire').mockReturnValue({} as any);
    await new Promise(r => setTimeout(r, 50));
});