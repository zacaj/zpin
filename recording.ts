import { Log } from './log';
import { Timer, passTime, setTime, time, wait } from './timer';
import * as fs from 'fs';
import { split, assert, debugging } from './util';
import { matrix } from './switch-matrix';
import { settleForks } from './promises';
import { MPU } from './mpu';
import { initMachine } from './init';
import { Events } from './events';
import { machine } from './machine';

export let curRecording: string|undefined = undefined;
let lines: string[];
let curTime = 0;
let curLine = 0;

export function initRecording(recording: string) {
    curRecording = recording;
    Log.log(['console'], 'playing back recording %s', recording);
    Timer.mockTime = 0;
    const file = fs.readFileSync(curRecording!, 'utf8');
    lines = file.split('\n');
    curTime = 0;
    curLine = 1;
}

export async function playRecording(toPoint?: string) {
    // await new Promise(r => setTimeout(r, 10000));
    const events = Events;
    const timer = Timer;
    const _machine = machine;
    for (; curLine < lines.length; curLine++) {
        const line = lines[curLine];
        if (line.startsWith('#')) continue;
        if (line && !line.includes(' ')) {
            const bp = line;
            if (bp === 'debug' && debugging()) {
                const nextLine = lines[curLine+1];
                debugger;
            } else if (process.env.NODE_ENV === 'test') {
                assert(bp === toPoint, `expected breakpoint ${toPoint} but got '${bp}'`);
                curLine++;
                await settleForks();
                return;
            }
            continue;
        }
        const [ts, sw, stateS, colS, rowS, timeS] = split(line, ' switch \'', '\' state -> ', ' (', ',', ') at ');
        if (timeS === undefined) continue;

        const time = parseInt(timeS, 10);
        const diff = time - curTime;
        assert(diff>=0, 'time travel not implemented');
        if (diff)
            await passTime(diff);
        curTime = time;

        matrix[parseInt(colS, 10)][parseInt(rowS, 10)]!.changeState(stateS === 'true', 'recording '+curRecording);
        await settleForks();
    }
    await settleForks();

    if (process.env.NODE_ENV !== 'test')
        await setTime(undefined);
    Log.log(['console', 'switch'], 'recording %s finished at %i', curRecording, time());
    curRecording = undefined;
}


export async function testRecording(name: string, toPoint?: string) {
    jest.spyOn(Log, 'init').mockImplementation(() => {});
    jest.spyOn(MPU, 'sendCommandCode').mockImplementation(async (cmd) => {
        return {
            code: 200,
            resp: 'mocked',
        };
    });
    await initMachine(false, false, true, false, `./recordings/${name}.rec`, toPoint);
}

export async function continueRecording(toPoint?: string) {
    await playRecording(toPoint);
}
export async function finishRecording() {
    await playRecording();
}
