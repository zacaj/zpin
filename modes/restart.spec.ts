import { machine } from '../machine';
import { testRecording, finishRecording } from '../recording';
import { passTime } from '../timer';
import { snapshot } from '../jest';

describe('restart', () => {
    test('light goes away', async () => {
        await testRecording('restartFailedPoker');
        expect(machine.out!.treeValues.lRampArrow).toEqual([]);
        snapshot();
    });
});