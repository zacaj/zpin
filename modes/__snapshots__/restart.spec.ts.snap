// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`restart light goes away 1`] = `
Object {
  "outputs": Object {
    "centerBank": false,
    "getSkillshot": [Function],
    "iCenter1": "8H",
    "iCenter2": "QS",
    "iCenter3": "QD",
    "iLeft1": "AH",
    "iLeft2": "JD",
    "iLeft3": "QH",
    "iLeft4": "7S",
    "iMini1": "5C",
    "iMini2": "7D",
    "iMini3": "10H",
    "iRight1": "KH",
    "iRight2": "7H",
    "iRight3": "JS",
    "iRight4": "5D",
    "iRight5": "6S",
    "iSS1": "",
    "iSS2": "",
    "iSS3": "",
    "iSS4": "",
    "iSS5": "",
    "iSS6": "",
    "iSpinner": "",
    "iUpper21": "JH",
    "iUpper22": "7C",
    "iUpper31": "10C",
    "iUpper32": "5H",
    "iUpper33": "9H",
    "ignoreSkillsot": Set {},
    "kickerEnable": true,
    "lEjectArrow": Array [],
    "lEjectShowCards": Array [],
    "lEjectStartMode": Array [],
    "lLaneUpper1": Array [
      "Orange",
    ],
    "lLaneUpper2": Array [
      "Orange",
    ],
    "lLaneUpper3": Array [
      "Orange",
    ],
    "lLaneUpper4": Array [
      "Orange",
    ],
    "lLeftArrow": Array [],
    "lLowerRamp": Array [],
    "lMagnaSaveStatus": Array [
      "Green",
    ],
    "lMainTargetArrow": Array [],
    "lMiniReady": Array [
      "Red",
    ],
    "lPopperStatus": Array [
      "Green",
    ],
    "lPower1": Array [
      "Orange",
    ],
    "lPower2": Array [
      "Orange",
    ],
    "lPower3": Array [],
    "lPower4": Array [],
    "lRampArrow": Array [],
    "lRampShowCards": Array [],
    "lRampStartMb": Array [],
    "lShooterLaneArrow": Array [],
    "lShooterShowCards": Array [],
    "lShooterStartHand": Array [],
    "lSideShotArrow": Array [],
    "lSideTargetArrow": Array [],
    "lSpinnerArrow": Array [],
    "lUpperLaneArrow": Array [],
    "lUpperTargetArrow": Array [],
    "leftBank": false,
    "leftGate": false,
    "leftMagnet": false,
    "lockPost": false,
    "magnetPost": false,
    "miniBank": false,
    "miniDiverter": false,
    "miniEject": false,
    "miniFlipperEnable": false,
    "outhole": false,
    "popper": false,
    "rampUp": true,
    "realRightBank": false,
    "right1": false,
    "right2": false,
    "right3": false,
    "right4": false,
    "right5": false,
    "rightBank": false,
    "rightGate": true,
    "shooterDiverter": true,
    "troughRelease": false,
    "upper2": false,
    "upper3": false,
    "upperEject": false,
    "upperMagnet": false,
  },
  "state": Object {
    "_name": "Machine10",
    "children": Array [
      Object {
        "_name": "DropBank11",
        "children": Array [],
        "name": "DropBank11",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank12",
        "children": Array [],
        "name": "DropBank12",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank13",
        "children": Array [],
        "name": "DropBank13",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank14",
        "children": Array [],
        "name": "DropBank14",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank15",
        "children": Array [],
        "name": "DropBank15",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank16",
        "children": Array [],
        "name": "DropBank16",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "Game20",
        "children": Array [
          Object {
            "_name": "Player21",
            "children": Array [
              Object {
                "_name": "ClearHoles22",
                "children": Array [],
                "name": "ClearHoles22",
                "ownOuts": Object {
                  "miniEject": false,
                  "outhole": false,
                  "upperEject": false,
                },
                "state": Object {},
              },
              Object {
                "_name": "Ball26",
                "children": Array [
                  Object {
                    "_name": "ResetAnyDropOnComplete27",
                    "children": Array [
                      Object {
                        "_name": "DropBankResetter28",
                        "children": Array [],
                        "name": "DropBankResetter28",
                        "ownOuts": Object {
                          "upper3": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter29",
                        "children": Array [],
                        "name": "DropBankResetter29",
                        "ownOuts": Object {
                          "upper2": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter30",
                        "children": Array [],
                        "name": "DropBankResetter30",
                        "ownOuts": Object {
                          "centerBank": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter31",
                        "children": Array [],
                        "name": "DropBankResetter31",
                        "ownOuts": Object {
                          "miniBank": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter32",
                        "children": Array [],
                        "name": "DropBankResetter32",
                        "ownOuts": Object {
                          "leftBank": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter33",
                        "children": Array [],
                        "name": "DropBankResetter33",
                        "ownOuts": Object {
                          "rightBank": false,
                        },
                        "state": Object {},
                      },
                    ],
                    "name": "ResetAnyDropOnComplete27",
                    "ownOuts": undefined,
                    "state": Object {},
                  },
                ],
                "name": "Ball26",
                "ownOuts": Object {},
                "state": Object {
                  "skillshot": undefined,
                },
              },
              Object {
                "_name": "Spinner23",
                "children": Array [],
                "name": "Spinner23",
                "ownOuts": Object {
                  "iSpinner": undefined,
                  "leftGate": false,
                },
                "state": Object {
                  "comboMult": 1,
                  "rounds": 0,
                  "score": 800,
                },
              },
              Object {
                "_name": "LeftOrbit24",
                "children": Array [],
                "name": "LeftOrbit24",
                "ownOuts": Object {
                  "rightGate": true,
                },
                "state": Object {
                  "comboMult": 1,
                  "rounds": 1,
                  "score": 20000,
                },
              },
              Object {
                "_name": "Poker43",
                "children": Array [],
                "name": "Poker43",
                "ownOuts": Object {
                  "getSkillshot": [Function],
                  "iCenter1": "8H",
                  "iCenter2": "QS",
                  "iCenter3": "QD",
                  "iLeft1": "AH",
                  "iLeft2": "JD",
                  "iLeft3": "QH",
                  "iLeft4": "7S",
                  "iMini1": "5C",
                  "iMini2": "7D",
                  "iMini3": "10H",
                  "iRight1": "KH",
                  "iRight2": "7H",
                  "iRight3": "JS",
                  "iRight4": "5D",
                  "iRight5": "6S",
                  "iUpper21": "JH",
                  "iUpper22": "7C",
                  "iUpper31": "10C",
                  "iUpper32": "5H",
                  "iUpper33": "9H",
                  "lEjectShowCards": Array [],
                  "lRampShowCards": Array [],
                  "lShooterShowCards": Array [],
                  "lockPost": undefined,
                  "rampUp": undefined,
                  "shooterDiverter": true,
                  "upperEject": undefined,
                },
                "state": Object {
                  "bet": 100,
                  "closeShooter": false,
                  "dealerCardsUsed": Array [],
                  "dealerHand": Array [
                    Object {
                      "flipped": true,
                      "num": 3,
                      "suit": "d",
                    },
                    Object {
                      "flipped": true,
                      "num": 6,
                      "suit": "h",
                    },
                    null,
                    null,
                    null,
                    null,
                    null,
                  ],
                  "dealerHandDesc": undefined,
                  "newMbs": Array [],
                  "newModes": Array [],
                  "playerCardsUsed": Array [],
                  "playerHand": Array [
                    Object {
                      "num": 3,
                      "suit": "h",
                    },
                    Object {
                      "num": 1,
                      "suit": "d",
                    },
                    null,
                    null,
                    null,
                    null,
                    null,
                  ],
                  "playerHandDesc": undefined,
                  "playerWins": undefined,
                  "pot": 400,
                  "slots": Array [
                    Object {
                      "num": 8,
                      "suit": "h",
                    },
                    Object {
                      "num": 12,
                      "suit": "s",
                    },
                    Object {
                      "num": 12,
                      "suit": "d",
                    },
                    Object {
                      "num": 13,
                      "suit": "h",
                    },
                    Object {
                      "num": 7,
                      "suit": "h",
                    },
                    Object {
                      "num": 11,
                      "suit": "s",
                    },
                    Object {
                      "num": 5,
                      "suit": "d",
                    },
                    Object {
                      "num": 6,
                      "suit": "s",
                    },
                    Object {
                      "num": 1,
                      "suit": "h",
                    },
                    Object {
                      "num": 11,
                      "suit": "d",
                    },
                    Object {
                      "num": 12,
                      "suit": "h",
                    },
                    Object {
                      "num": 7,
                      "suit": "s",
                    },
                    Object {
                      "num": 11,
                      "suit": "h",
                    },
                    Object {
                      "num": 7,
                      "suit": "c",
                    },
                    Object {
                      "num": 10,
                      "suit": "c",
                    },
                    Object {
                      "num": 5,
                      "suit": "h",
                    },
                    Object {
                      "num": 9,
                      "suit": "h",
                    },
                    Object {
                      "num": 5,
                      "suit": "c",
                    },
                    Object {
                      "num": 7,
                      "suit": "d",
                    },
                    Object {
                      "num": 10,
                      "suit": "h",
                    },
                  ],
                  "step": 2,
                },
              },
              Object {
                "_name": "PlayerOverrides25",
                "children": Array [],
                "name": "PlayerOverrides25",
                "ownOuts": Object {
                  "leftGate": undefined,
                  "rightGate": undefined,
                  "shooterDiverter": undefined,
                },
                "state": Object {},
              },
            ],
            "name": "Player21",
            "ownOuts": Object {
              "lEjectStartMode": Array [],
              "lLaneUpper1": Array [
                "Orange",
              ],
              "lLaneUpper2": Array [
                "Orange",
              ],
              "lLaneUpper3": Array [
                "Orange",
              ],
              "lLaneUpper4": Array [
                "Orange",
              ],
              "lMagnaSaveStatus": Array [
                "Green",
              ],
              "lMiniReady": Array [
                "Red",
              ],
              "lPopperStatus": Array [
                "Green",
              ],
              "lPower1": Array [
                "Orange",
              ],
              "lPower2": Array [
                "Orange",
              ],
              "lPower3": Array [],
              "lPower4": Array [],
              "lRampStartMb": Array [],
              "lShooterStartHand": Array [],
              "leftMagnet": false,
              "rampUp": true,
              "shooterDiverter": undefined,
            },
            "state": Object {
              "chips": 2,
              "closeShooter": false,
              "focus": "Poker43",
              "laneChips": Array [
                true,
                true,
                true,
                true,
              ],
              "mbsQualified": Array [],
              "miniReady": false,
              "modesQualified": Array [],
              "rampUp": true,
              "score": 106020,
            },
          },
        ],
        "name": "Game20",
        "ownOuts": Object {
          "kickerEnable": true,
        },
        "state": Object {
          "ballNum": 1,
          "playerUp": 0,
          "players": Array [
            "Player21",
          ],
        },
      },
      Object {
        "_name": "MachineOverrides17",
        "children": Array [
          Object {
            "_name": "EosPulse18",
            "children": Array [],
            "name": "EosPulse18",
            "ownOuts": undefined,
            "state": Object {},
          },
        ],
        "name": "MachineOverrides17",
        "ownOuts": Object {
          "rampUp": undefined,
          "realRightBank": false,
          "shooterDiverter": undefined,
        },
        "state": Object {},
      },
    ],
    "name": "Machine10",
    "ownOuts": Object {
      "centerBank": false,
      "getSkillshot": undefined,
      "iCenter1": "",
      "iCenter2": "",
      "iCenter3": "",
      "iLeft1": "",
      "iLeft2": "",
      "iLeft3": "",
      "iLeft4": "",
      "iMini1": "",
      "iMini2": "",
      "iMini3": "",
      "iRight1": "",
      "iRight2": "",
      "iRight3": "",
      "iRight4": "",
      "iRight5": "",
      "iSS1": "",
      "iSS2": "",
      "iSS3": "",
      "iSS4": "",
      "iSS5": "",
      "iSS6": "",
      "iSpinner": "",
      "iUpper21": "",
      "iUpper22": "",
      "iUpper31": "",
      "iUpper32": "",
      "iUpper33": "",
      "ignoreSkillsot": Set {},
      "kickerEnable": false,
      "lEjectArrow": Array [],
      "lEjectShowCards": Array [],
      "lEjectStartMode": Array [],
      "lLaneUpper1": Array [],
      "lLaneUpper2": Array [],
      "lLaneUpper3": Array [],
      "lLaneUpper4": Array [],
      "lLeftArrow": Array [],
      "lLowerRamp": Array [],
      "lMagnaSaveStatus": Array [],
      "lMainTargetArrow": Array [],
      "lMiniReady": Array [
        "Red",
      ],
      "lPopperStatus": Array [],
      "lPower1": Array [],
      "lPower2": Array [],
      "lPower3": Array [],
      "lPower4": Array [],
      "lRampArrow": Array [],
      "lRampShowCards": Array [],
      "lRampStartMb": Array [],
      "lShooterLaneArrow": Array [],
      "lShooterShowCards": Array [],
      "lShooterStartHand": Array [],
      "lSideShotArrow": Array [],
      "lSideTargetArrow": Array [],
      "lSpinnerArrow": Array [],
      "lUpperLaneArrow": Array [],
      "lUpperTargetArrow": Array [],
      "leftBank": false,
      "leftGate": false,
      "leftMagnet": false,
      "lockPost": false,
      "magnetPost": false,
      "miniBank": false,
      "miniDiverter": false,
      "miniEject": false,
      "miniFlipperEnable": false,
      "outhole": false,
      "popper": false,
      "rampUp": false,
      "realRightBank": false,
      "right1": false,
      "right2": false,
      "right3": false,
      "right4": false,
      "right5": false,
      "rightBank": false,
      "rightGate": false,
      "shooterDiverter": false,
      "troughRelease": false,
      "upper2": false,
      "upper3": false,
      "upperEject": false,
      "upperMagnet": false,
    },
    "state": Object {
      "ballsInTrough": 4,
      "ballsLocked": 0,
      "lastSwitchHit": Switch {
        "_state": false,
        "column": 3,
        "lastChange": 95065,
        "lastClosed": 94805,
        "lastOpened": 95065,
        "minOffTime": 50,
        "minOnTime": 0,
        "name": "single standup",
        "row": 7,
      },
      "lockDown": false,
      "miniDown": false,
    },
  },
}
`;
